### ISSUE TITLE
Life360 JSON question

### ISSUE THREAD

**Issue by hepcat72** [kaylathedev/life360-node-api#2]
It looks like you've done some dissection of the Life360 JSON data.  I wonder if you could provide any insights to what I'm seeing.  I forked node-red-contrib-life and noted that the location node's code checks for changes to a member's named location by looking at the location name every 15 seconds for each member.  My node red instance was outputting location change events at somewhat frequent intervals when no change to the location name had actually occurred.  I figured out the reason for that, but I also discovered something else:

The code which checks for location name changes detects a change when the location name goes from a string to null or vice versa.  It turns out that every once in awhile, even though a person's location hasn't changed, the name will randomly be null.  There's never any output from the node in those instances, but the tracking code definitely records it as a change.  It must just be filtered out later in the code since the node isn't triggered to output.

So I was looking for a way to not update that tracking data in these instances, but preserve the ability to detect when someone actually crosses a geofence boundary (where "null" actually means something).  Looking at the member JSON data and comparing these things, 2 differences jumped out at me:

`member.location.source` and `member.location.sourceId` are both null when the name is null and not null when the name is not null.  Without going for a drive, I don't know if they're null when I'm not at a named place, so I wondered if you have learned what source and sourceId are for and how they're populated.

**Comment by kaylathedev** [comment]
I haven't worked with Life360's api enough to actually see differences in the responses. But I'm going to work on my code over the next week, while using a location faker app, to see if I can see what you're seeing.

If I'm understanding you correctly, you have already added a Place in Life360. And when you're at that Place, the api at "www.life360.com" will sometimes set the `member.location.name` field to null, and other times, it'll be the name of the place?

That's interesting, because I would think Life360 should always set `member.location.name` to be equal to the place name, if you're inside the geofenced boundary.

**Comment by kaylathedev** [comment]
Also, I don't understand what the `source` and `sourceId` fields are. During my testing, `member.location.source` is always the letter `l`, and `member.location.sourceId` is formatted like a GUID.

**Comment by hepcat72** [comment]
Actually, I've learned a lot since that question.  I'm trying to add features to node-red-contrib-life and it turns out that that package had some bugs.  The nulls I was seeing were due to the same member being in multiple circles where either the circles had the same place with a different name or no place.  The nulls were coming from the member record in the circle where that place didn't exist.  And the thing that made it confusing was the random order in which the circles were processed, which is what had made me think that the null place names were "occasional".

I have fixed those issues in my fork and now it appears that the place name is in fact reliable and consistent.

Now that I'm past the bugs, I'm learning about javascript and client/server communications.  The amount I don't know is vast.  I'm currently trying to figure out how to resolve a javascript "Promise" and I don't understand why I'm not getting a resolved JSON object...  So I guess I've jumped into the deep end without knowing how to swim.  ;)

**Comment by kaylathedev** [comment]
Glad you fixed the issue with the null place names!

To address your other problem, (I assume you mean you already have a promise, and you want to use the returned json object). You have 2 ways of getting the json object from promises.

Method # 1: use the `await` keyword before the promise. So let's say you have a method named `life360.login` and that method returns a `Promise` object. Then writing `var json = await life360.login();` will get the result from the Promise.

Method # 2: Attach a callback to the promise by calling `.then(callback)`. For example, `life360.login().then(onLoginSuccessful);`, where `onLoginSuccessful` is a function that receives the json as an argument.

You could also inline the callback function. `life360.login( function (json) { /* do something with the json result */ } );

See [stackoverflow.com/questions/29516390/how-to-access-the-value-of-a-promise](https://stackoverflow.com/questions/29516390/how-to-access-the-value-of-a-promise)

If I was wrong, you created a promise, and you don't know how to resolve the promise, then see this. [freecodecamp.org/news/javascript-es6-promises-for-beginners-resolve-reject-and-chaining-explained/#creating-and-using-a-promise-step-by-step](https://www.freecodecamp.org/news/javascript-es6-promises-for-beginners-resolve-reject-and-chaining-explained/#creating-and-using-a-promise-step-by-step). Scroll down to "Creating and Using A Promise Step by Step".

If you have any questions, feel free to comment here, or send me an email. I'm more than happy to share what I've learned!

**Comment by hepcat72** [comment]
Thanks! There's some sort of nuance that seemed to prevent me from making those strategies work, or rather, they worked, but not in the place I needed them to work.  (Take everything I say with a grain of salt, BTW.  My understanding and/or attempts may be incomplete/flawed.)

For some reason, I woke up this morning at around 3am and I ended up banging my head against this until I got it to work (albeit somewhat hackily/buggily).  Here's what I learned about this specific case.

My attempts at using `await` were failing because the function from which I was calling the data retrieval method (`RED.httpAdmin.get()` - called from a function inside `oneditprepare` in the node's html file) was not an async function, so I couldn't use `await` there.  I could call async functions containing `await` and the functions I passed in had access to the data, but I had no way to get it back to `RED.httpAdmin.get()` (or the anonymous function(req,res) in its parameters where the parsing of the data was happening to pass it back to the oneditprepare function that created the select list from it).

I also tried the callback strategy multiple times in different ways and I never ended up with anything other than promise json, undefined, or (weirdly) `{}` in the `RED.httpAdmin.get()` (or its function parameter's(?)) code.

I eventually got something that works (sometimes).  It was a huge relief and I imagine it will lead to a stable solution.  I ended up using await in an anonymous async function followed by a sleep while that function worked.  The key was I had to pass in the node and set a data member in the node's instance(?), which I then accessed after the sleep.

I suspect that one of my other cleaner attempts may work as well, but I don't know how to avoid the old-style sleep - but maybe I don't even need to do the sleep.  I didn't play with it any further once I saw it work one time and decided to stop while I was ahead.  I'll play with it some more later.

It actually inspired me to turn it around and save the circle data in the server node when it first connects.  That way, I wouldn't need to do any of this promise stuff at all.  I can just directly access the data, assuming that the server has initialized.
